### PR TITLE
Add support for "private_key_jwt" client authentication method

### LINKED ISSUE #2065: Client assertion (private_key_jwt) support
Hey,

I've spent some time setting up TLS certificates for client authentication and was completely disappointed with the complexity of the setup. However, I really needed some asymmetric key solution for my setup. And I wanted it to be as _standard_ as it gets.

So I've implemented the necessary extensions for the IdentityServer to add support for _private_key_jwt_ (RS256 signed tokens) client authentication method as defined in the [OIDC spec](http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication)
Still need to add some extra test coverage for the edge cases, but otherwise it is a pretty straightforward implementation.
And I didn't bother extending the metadata endpoint to indicate support for it so far (since it is kinda hardcoded and ain't easy to do from the outside).

There are no changes to the configuration necessary in the simplest case.

Here are some highlights:

- Implementation is basically about implementing the necessary ISecretParser/ISecretValidator pair.
- I reuse same existing `X509CertificateThumbprint` client secrets used for client certificate authentication.
- If client assertion token embeds `x5c` certificate key, its thumbprint is checked against client secrets and if it matches, I use it to validate the token.
- To validate tokens that only embed `x5t`, users can provide an implementation of `IClientCertificateStore`. Can add a default implementation that would use "Other People" certificate store or something like that (would appreciate an advice on this).
- Token replay protection is supported
- Tested for compatibility with at least one Java client implementation: [connect2id](http://connect2id.com/products/nimbus-oauth-openid-connect-sdk)

_Note: It should be fairly easy to add client_secret_jwt (HMAC signed tokens) method support as well, but the way I see it, it would require storing shared keys unhashed, which requires some extra consideration first, IMO._

So now my question is: would you like to have a pull request for this?

### PR DESCRIPTION
**Pull request by uncleyo**
This authentication method is defined in OpenId Connect 1.0 specification.
It enables simpler authentication method based on asymmetric keys compared to TLS client certificates.
Initially only "x5t" and "x5c" keys are supported, but other JWK key type support could be added later.

Closes #2065
Supersedes #2144

### CODE REVIEW

**Comment by uncleyo** [comment]
I have rebased to the latest dev branch version and fixed what got broken.

Two observations that probably deserve separate PRs:
- Every secret validator has to check secrets expiration. How about filtering them before passing into validators?
- IssuerUri/DynamicallyCalculatedIssuerUri seem to be inversed. Shouldn't it be that IssuerUri defaults to DynamicallyCalculatedIssuerUri value?

**Comment by leastprivilege** [comment]
thanks again. still swamped with other work. but will look into it asap.

**Comment by leastprivilege** [comment]
could you update again from current dev?

**Comment by leastprivilege** [comment]
Finally got some time to look at this - why do we need the replay protection cache?

**Comment by uncleyo** [comment]
Not sure I understand.

Do you mean
- why do we need the replay protection cache at all
 * https://tools.ietf.org/html/rfc7523#section-6 says it's not mandatory, but I thought it would be good to do so
- why expose it in the api
 * to let people use some farm-friendly cache
- why a new cache interface and not existing ICache?
 * because existing cache lacks `bool Add(...)` semantics (and writes to debug on cache miss, making it noisy)

We could remove it from the api and just use in-memory implementation if you want to keep things easy on the outside. How about that?

Did at least one of those points answer your question? :)

**Comment by leastprivilege** [comment]
I meant

* why do we need the replay protection cache at all
https://tools.ietf.org/html/rfc7523#section-6 says it's not mandatory, but I thought it would be good to do so

I never really read this spec so you are much deeper into the topic than I am - but in essence the signed JWT becomes a "bearer" token (comparable to a shared secret - just without transmitting the actual secret over the wire. right? 

So I am confused why a replay cache is needed? or is every JWT somehow unique (maybe includes a nonce or jit?)

**Comment by uncleyo** [comment]
Yep, the 'jti" is used as nonce, and it is optional according to https://tools.ietf.org/html/rfc7523, but not according to http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication.

**Comment by leastprivilege** [comment]
OK - got it. need to read that spec ;)

Lemme think about the cache problem.

**Comment by leastprivilege** [comment]
I think i would prefer not to add another cache right now - and only add the validator/parser.

We could add a better cache in the next major release.

**Comment by uncleyo** [comment]
Ok, even internally or just on the outside?

**Comment by leastprivilege** [comment]
yes - let's ony add the validator and parser (and its necessary dependencies) for now.

do you want to do a new PR - the current one does not merge anymore.

**Comment by uncleyo** [comment]
I'll make the changes.
Do you know if it is ok to rebase branches that a pull-request is pointing at on GitHub? If not, I'll simply recreate the PR.

**Comment by leastprivilege** [comment]
don't know ;)

**Comment by uncleyo** [comment]
Now we know that it's supported :)

File: source/Core/Validation/ClientAssertionSecretParser.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        /// <param name="environment">The environment.</param>
+        /// <returns>
+        /// A parsed secret
+        /// </returns>
+        public async Task<ParsedSecret> ParseAsync(IDictionary<string, object> environment)
+        {
+            Logger.Debug("Start parsing for JWT client assertion in post body");
+
+            var context = new OwinContext(environment);
+            var body = await context.Request.ReadFormAsync();
```
Comment (leastprivilege) [inline source/Core/Validation/ClientAssertionSecretParser.cs]: Could you please use

`context.ReadRequestFormAsync` instead.

**Comment by leastprivilege** [comment]
Also  - could you for now remove the parser and validator from the collection. and also from the disco doc.

For the start i want to make it optional but add a sample for it.

I am also working on a better discovery mechanism for authN methods.

**Comment by uncleyo** [comment]
Done. Rebased again.

File: source/Core/Extensions/JwtSecurityTokenExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+namespace IdentityServer3.Core.Extensions
+{
+    internal static class JwtSecurityTokenExtensions
+    {
+        public static X509Certificate2 GetCertificateFromToken(this JwtSecurityToken securityToken)
+        {
+            object values;
+            if (securityToken.Header.TryGetValue(JwtHeaderParameterNames.X5c, out values))
+            {
+                var rawCertificate = (values as object[])?.Cast<string>().FirstOrDefault();
```
Comment (leastprivilege) [inline source/Core/Extensions/JwtSecurityTokenExtensions.cs]: No c# 6 syntax please

File: source/Core/Validation/ClientAssertionSecretParser.cs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        public async Task<ParsedSecret> ParseAsync(IDictionary<string, object> environment)
+        {
+            Logger.Debug("Start parsing for JWT client assertion in post body");
+
+            var context = new OwinContext(environment);
+            var body = await context.ReadRequestFormAsync();
+
+            if (body != null)
+            {
+                var clientId = body.Get(Constants.TokenRequest.ClientId);
```
Comment (leastprivilege) [inline source/Core/Validation/ClientAssertionSecretParser.cs]: According to spec the client id is the sub claim of the assertion

Comment (uncleyo) [inline source/Core/Validation/ClientAssertionSecretParser.cs]: That is accounted for in GetClientIdFromToken below, but the example [here](http://openid.net/specs/openid-connect-core-1_0.html#ClientAuthentication) also includes it in the post body.
So in this implementation the one in body is optional and the one in the claim is enforced [here](https://github.com/uncleyo/IdentityServer3/blob/private_key_jwt/source/Core/Validation/PrivateKeyJwtSecretValidator.cs#L114).

File: source/Core/Extensions/JwtSecurityTokenExtensions.cs (hunk 2 above)
Comment (uncleyo) [inline source/Core/Extensions/JwtSecurityTokenExtensions.cs]: No problem, I'll update it. Is it because you still support VS 2013 in IdSrv3?

Comment (leastprivilege) [inline source/Core/Extensions/JwtSecurityTokenExtensions.cs]: yes

Comment (leastprivilege) [inline source/Core/Extensions/JwtSecurityTokenExtensions.cs]: OK - so make this final change and we are good to go.

**Comment by uncleyo** [comment]
Done.

File: source/Core/Validation/PrivateKeyJwtSecretValidator.cs [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+                && enumeratedSecrets.Any(s => s.Type == Constants.SecretTypes.X509CertificateThumbprint))
+            {
+                Logger.Warn("Cannot validate client assertion token that does not embed full certificate using only thumbprint secret");
+                return fail;
+            }
+
+            var tokenValidationParameters = new TokenValidationParameters
+            {
+                IssuerSigningKey = embeddedKey,
+                IssuerSigningKeys = savedKeys,
```
Comment (leastprivilege) [inline source/Core/Validation/PrivateKeyJwtSecretValidator.cs]: Why do you set both?

Comment (uncleyo) [inline source/Core/Validation/PrivateKeyJwtSecretValidator.cs]: Only one is ever used, but I think I know how to improve it. Stay tuned.

**Comment by uncleyo** [comment]
There

**Comment by leastprivilege** [comment]
Hey, 

i just did some refactoring on the parser/validator plumbing - e.g. moved the expiration check out - 

have a look here

https://github.com/IdentityServer/IdentityServer3/pull/2310

feel free to review - and once I merged you have to adjust your PR again (sorry)

**Comment by uncleyo** [comment]
That's ok. Seems like I'll only have to remove the expiration check.

**Comment by leastprivilege** [comment]
you need to merge again.

**Comment by uncleyo** [comment]
Ding :)

**Comment by leastprivilege** [comment]
thanks! good job!

**Comment by uncleyo** [comment]
Thanks!
Second thought: how about adding support for `X509CertificateBase64` for SSL client certs (adding another `X509CertificateBase64SecretValidator.cs`) for symmetry? I can make a PR.

Also, please tell me what you think about https://github.com/IdentityModel/IdentityModel/pull/28.

**Comment by leastprivilege** [comment]
it might make sense to re-design the client secret plumbing in the TokenClient altogether (and maybe something that can be shared amongst all the client libs). This is a v2 thing.

Open a new issue on the IdentityModel repo so we can start a discussion. But I am currently swamped with other work.
